packages/kit/src/runtime/client/client.js

```diff
@@ -76,7 +78,7 @@ if (DEV && BROWSER) {
 		if (!stack) return;
 		if (!stack[0].includes('https:') && !stack[0].includes('http:')) stack = stack.slice(1); // Chrome includes the error message in the stack
 		stack = stack.slice(2); // remove `warn` and the place where `warn` was called
-		if (stack[0].includes(import.meta.url)) return;
+		if (stack[0]?.includes(current_module_url)) return;
```

Comment:
What's this about? Should we instead do `!stack[0] || ..` because if there's no stack how can we know it's not a false positive?

Reply:
That sounds more desirable. Initially, I was running into an edge case where Safari doesn’t print the full stack like Chrome does, so after doing `stack = stack.slice(2)` there’s nothing left in the array.

Chrome’s stack looks something like this when clicking a button that does `history.pushState()`:
0: error message
1: warn()
2: where warn was called
3: the event listener
4: the function that handles event listeners(?)

While Safari’s stack is just:
0: warn()
1: where warn was called

Reply:
Thinking about it more, it should be `stack[0] && ...` because if the array is empty after removing the first two items in the stack, then `history.pushState` was called internally by Kit.

Reply:
But isn't the place where warn is called our monkey patched replaceState/pushState? That one should be ignored, and if there's nothing after it then we cannot know whether or not it was called correctly and not warn. @Rich-Harris do you see the same behavior with the safari call stack?

Reply:
I went back and did more testing (this time actually recording my results). I think I was wrong and confused it being a Safari behaviour when it's probably because it's from an anonymous function?

### Anonymous function without a body

This is the call stack for calling `history.pushState()` in the template as a single line anonymous function aka `onclick={() => history.pushState()`. We can fire a warning if the stack array is empty (but maybe that's not definitive?)

```js
0: warn@http://localhost:5173/node_modules/.pnpm/@sve…/@sveltejs/kit/src/runtime/client/client.js:75:24
1: @http://localhost:5173/node_modules/.pnpm/@sveltej…s/@sveltejs/kit/src/runtime/client/client.js:93:7
```

### Anonymous / named function with a body

This is the call stack for calling `history.pushState()` in a svelte component in a named function (in script tag or in the html template, etc.). We're able to detect that it comes from userland because of the svelte file and so we fire a warning.
```
0: warn@http://localhost:5173/node_modules/.pnpm/@sve…/@sveltejs/kit/src/runtime/client/client.js:75:24
1: @http://localhost:5173/node_modules/.pnpm/@sveltej…s/@sveltejs/kit/src/runtime/client/client.js:93:7
2: onClick@http://localhost:5173/src/routes/+layout.svelte:246:19
```

### SvelteKit internals / navigation APIs

This is an example call stack for kit event listeners and APIs like `goto` that are called from a svelte file. The svelte file is omitted from the call stack if it's called inline like example 1
```
0: warn@http://localhost:5173/node_modules/.pnpm/@sve…/@sveltejs/kit/src/runtime/client/client.js:75:24
1: @http://localhost:5173/node_modules/.pnpm/@sveltej…s/@sveltejs/kit/src/runtime/client/client.js:99:7
2: @http://localhost:5173/node_modules/.pnpm/@sveltej…sveltejs/kit/src/runtime/client/client.js:2191:24
```

Reply:
Ooh I see! Yeah this makes sense then! Can you put a small comment above the check explaining that no third stack entry means a direct anonymous function call and therefore should warn? After that this is good to go - thanks for the in-depth exploration and explanation!

Reply:
```suggestion
		// Can be falsy if was called directly from an anonymous function
		if (stack[0]?.includes(current_module_url)) return;
```

Reply:
> Ooh I see! Yeah this makes sense then! Can you put a small comment above the check explaining that no third stack entry means a direct anonymous function call and therefore should warn? After that this is good to go - thanks for the in-depth exploration and explanation!

Sorry I didn't get around to this yesterday. I saw it on my phone but forgot to do it when I got back home.

Reply:
No worries, just did the change, will merge soon